R/data.table.R

```diff
@@ -1877,75 +1877,181 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
         if (is.na(rownames) || rownames<1L || rownames>ncol(x))
           stop("as.integer(rownames)==", rownames,
                " which is outside the column number range [1,ncol=", ncol(x), "].")
+        if (length(x[[rownames]]) != nrow(x)) 
+          stop("x[,", rownames, "] has multi-column type (such as a matrix column) ",
+               "and cannot be used as rownames: length(x[,", rownames,
+               "])==", length(X[[rownames]]), "but should be nrow(x)==", nrow(x))
+        if (is.list(x[[rownames]]))
+          stop("x[,", rownames, "] is a list column, the rownames column should be a vector")
```

Comment:
Not sure if we want to error, warn, or just let the user use a list column as the rownames. The `rownames()` function when applied to a matrix will accept a list without warning or error, converting the list elements to a character vector (it looks like through dput).

Reply:
Downgraded to warning